ouroboros-consensus/src/unstable-consensus-testlib/Test/Ouroboros/Consensus/ChainGenerator/Adversarial.hs

```diff
@@ -249,20 +252,44 @@ checkAdversarialChain recipe adv = do
                 Nothing     -> pure ()   -- there are no remaining honest active slots
                                          --
                                          -- TODO hpc shows this never executes
-    checkDensity :: Exn.Except (AdversarialViolation hon adv) ()
-    checkDensity =
-        when (C.Count s <= C.windowSize winA) $ do
-            let -- first stability window after the intersection in the honest
-                -- schema
-                hScgWin = C.UnsafeContains (C.Count $ C.getCount $ C.windowStart winA) (C.Count s)
-                -- first stability window after the intersection in the
-                -- alternative schema
-                aScgWin = C.UnsafeContains (C.Count 0) (C.Count s)
-                -- number of active slots in the first stability window after
-                -- the intersection in the honest schema
-                hCount = C.getCount $ BV.countActivesInV S.notInverted (C.sliceV hScgWin vH)
-                aCount = C.getCount $ BV.countActivesInV S.notInverted (C.sliceV aScgWin vA)
-            when (aCount >= hCount) $ Exn.throwError $ BadDensity hCount aCount
+
+    -- | Check that the density of the adversarial schema is less than the
+    -- density of the honest schema in the first stability window after the
+    -- intersection and in any subwindow that contains the first race to the
+    -- k+1st block.
+    --
+    -- See description of @ensureLowerDensityInWindows@
+    checkDensity = do
+        let
+          -- window of the honest schema after the intersection
+          winHAfterIntersection =
+              C.UnsafeContains
+                  (fromJust $ C.toWindow winH $ C.windowStart winA)
+                  (C.Count $ C.getCount (C.windowSize winA))
+          -- honest schema after the intersection
+          vHAfterIntersection = C.sliceV winHAfterIntersection vH
+          iterH :: RI.Race adv
+          iterH =
+              maybe (RI.initConservative vHAfterIntersection) id
+            $ RI.init (Kcp k) vHAfterIntersection
+
+        -- first race window after the intersection
+        C.SomeWindow _ w0 <- let RI.Race x = iterH in pure x
+
+        let
+          w0' = C.UnsafeContains (C.windowStart w0) (min (C.Count s) (C.windowSize w0))
+          vvH = C.getVector vHAfterIntersection
```

Comment:
My initial instinct is that functions on `Vector`s should be defined in the `Test.Ouroboros.Consensus.ChainGenerator.BitVector` module, instead of via `C.getVector`.

Although there are several windows in play here, so it might not be trivial to choose a helpful type for such a thing. But at least what you have here could be transplanted there quite directly, and given strongly-typed domains and perhaps loosely-typed codomains.